src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,381 @@
+from abc import abstractmethod
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+from area_detector_handlers import HandlerBase
+from event_model import DocumentRouter, RunRouter
+from event_model.documents import StreamDatum, StreamResource
+from pydantic.utils import deep_update
+from tiled.client import from_profile, from_uri
+from tiled.client.utils import MSGPACK_MIME_TYPE, handle_error
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+from tiled.utils import safe_json_dump
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+    "AD_TIFF": "multipart/related;type=image/tiff",
+}
+DTYPE_LOOKUP = {"number": "<f8", "array": "<f8", "boolean": "bool", "string": "str", "integer": "int"}
+
+
+# TODO (probably in a future PR?):
+# Move these into external repo (probably area-detector-handlers)
+# and use the existing handler discovery mechanism.
+
+"""
+A Handler consumes documents from Bluesky. It composes details (DataSource and its Assets) that will go into the
+Tiled database. Each Handler (more specifically, a StreamHandler) is instantiated per a Stream Resource.
+
+Tiled Adapters will later use this to read the data, with good random access and bulk access support.
+
+We put this code into handlers so that additional, possibly very unusual, formats can be supported by users without
+getting a PR merged into Bluesky or Tiled.
+
+The STREAM_HANDLER_REGISTRY (see example below) and the Tiled catalog paramter adapters_by_mimetype can together be
+used to support:
+    - ingesting a new mimetype from Bluesky documents and generating DataSource and Asset with appropriate
+      parameters (the handler's job);
+    - Interpreting those DataSource and Asset parameters to do I/O (the adapter's job).
+"""
+
+# The below is confusing (to Dan).
+# We have a use for StreamDatum(s) in, data out.
+# No dependency on Tiled, its database, HTTP, etc.
+# We just want fast incremental access to streaming data.
+
+# In common:
+# - knowledge of format details (meaning of parameters like swmr)
+# - instance state to track things like num_total_rows.
+
+# Different:
+# - The "at rest" use case has a conceptual dependency on
+#   DataSource Asset.
+# - The streaming data use case has an import dependency o
+#   whatever the I/O library is (e.g. h5py).
+
+
+class Thing:
```

Comment:
We should remember to drop this sketch class.

Reply:
Sure! It was just a reminder for me to think about the IO.